cluster/src/Cardano/Cluster/Environment.hs

```diff
@@ -151,30 +152,44 @@ prepareEnvironment node@(NodeName nodeIdT, nodeType) nodes stateDir = runState $
     nodeId :: String
     nodeId = T.unpack nodeIdT
 
+    cIndex :: Int
+    cIndex = unsafeElemIndex node nodes
+
+    (nodeNames :: [NodeName], nodeTypes :: [NodeType]) = unzip nodes
+
+    nodeAddrs :: Env -> [(Maybe NetworkAddress, NetworkAddress)]
+    nodeAddrs env = flip evalState (0, 0, 0) $ forM nodeTypes $ \typ -> do
+        (c, r, w) <- get
+        case typ of
+            NodeCore ->
+                put (c + 1, r, w + 1) >> return (Just $ nextNtwrkAddr c addr, nextNtwrkAddr w waddr)
+            NodeRelay ->
+                put (c, r + 1, w + 1) >> return (Just $ nextNtwrkAddr (r + 100) addr, nextNtwrkAddr w waddr)
+            NodeEdge ->
+                put (c, r, w + 1) >> return (Nothing, nextNtwrkAddr w waddr)
+      where
+        -- NOTE Safe when called after 'withDefaultEnvironment'
+        addr :: NetworkAddress
+        addr = unsafeNetworkAddressFromString (env ! "LISTEN")
+
+        -- NOTE Safe when called after 'withDefaultEnvironment'
+        waddr :: NetworkAddress
+        waddr = unsafeNetworkAddressFromString (env ! "NODE_API_ADDRESS")
+
```

Comment:
Basically extracted this from `prepareTopology` because it's actually required in `prepareTLS` as well to construct corresponding manager.